channeldb/migration23/migration.go

```diff
@@ -0,0 +1,128 @@
+	// failure information, if any.
+	oldFailInfoKey = []byte("htlc-fail-info")
+
+	// htlcAttemptInfoKey is the key used as the prefix of an HTLC attempt
+	// to store the info about the attempt that was done for the HTLC in
+	// question. The HTLC attempt ID is concatenated at the end.
+	htlcAttemptInfoKey = []byte("ai")
+
+	// htlcSettleInfoKey is the key used as the prefix of an HTLC attempt
+	// settle info, if any. The HTLC attempt ID is concatenated at the end.
+	htlcSettleInfoKey = []byte("si")
+
+	// htlcFailInfoKey is the key used as the prefix of an HTLC attempt
+	// failure information, if any.The  HTLC attempt ID is concatenated at
+	// the end.
+	htlcFailInfoKey = []byte("fi")
+)
+
+// htlcBucketKey creates a composite key from prefix and id where the result is
+// simply the two concatenated. This is the exact copy from payments.go.
+func htlcBucketKey(prefix, id []byte) []byte {
+	key := make([]byte, len(prefix)+len(id))
+	copy(key, prefix)
+	copy(key[len(prefix):], id)
+	return key
+}
+
+// MigrateHtlcAttempts will gather all htlc-attempt-info's, htlcs-settle-info's
+// and htlcs-fail-info's from the attempt ID buckes and re-store them using the
+// flattened keys to each payment's payment-htlcs-bucket.
+func MigrateHtlcAttempts(tx kvdb.RwTx) error {
+	paymentsBucket := tx.ReadWriteBucket(paymentsRootBucket)
+	if paymentsBucket == nil {
+		return nil
+	}
+
+	return paymentsBucket.ForEach(func(k, v []byte) error {
+		// Get the bucket which contains the payment, fail if the key
+		// does not have a bucket.
+		bucket := paymentsBucket.NestedReadWriteBucket(k)
+		if bucket == nil {
+			return fmt.Errorf("non bucket element in '%v' bucket",
+				paymentsRootBucket)
+		}
+
+		htlcsBucket := bucket.NestedReadWriteBucket(paymentHtlcsBucket)
+		if htlcsBucket == nil {
+			// Nothing to migrate for this payment.
+			return nil
+		}
+
+		return migrateHtlcsBucket(htlcsBucket)
+	})
+}
+
+// migrateHtlcsBucket is a helper to gather, transform and re-store htlc attempt
+// key/values.
+func migrateHtlcsBucket(htlcsBucket kvdb.RwBucket) error {
+	kvs := make(map[string][]byte)
+
+	if err := htlcsBucket.ForEach(func(aid, v []byte) error {
+		attemptBucket := htlcsBucket.NestedReadWriteBucket(aid)
+
+		if attemptBucket == nil {
+			return fmt.Errorf("non bucket element in '%v' bucket",
+				paymentHtlcsBucket)
+		}
+
+		// Collect attempt/settle/fail infos.
+		attemptInfo := attemptBucket.Get(oldAttemptInfoKey)
+		if attemptInfo != nil {
+			key := string(htlcBucketKey(htlcAttemptInfoKey, aid))
+			kvs[key] = attemptInfo
+		}
+
+		settleInfo := attemptBucket.Get(oldSettleInfoKey)
+		if settleInfo != nil {
+			key := string(htlcBucketKey(htlcSettleInfoKey, aid))
+			kvs[key] = settleInfo
+		}
+
+		failInfo := attemptBucket.Get(oldFailInfoKey)
+		if failInfo != nil {
+			key := string(htlcBucketKey(htlcFailInfoKey, aid))
+			kvs[key] = failInfo
+		}
+
+		// Delete the attempt bucket.
+		return htlcsBucket.DeleteNestedBucket(aid)
```

Comment:
This all happens in one huge write transaction. Therefore the order here doesn't play a huge role, if anything goes wrong, the whole transaction is rolled back.
But that is actually the main concern. Because we need every migration to be capable of running atomically, we'll end up migrating _all_ payments in one transaction. For some nodes that might be a huge chunk of data.

Not sure what we can realistically do about it here...
I only have two ideas:
 1. Just go with it and hope it will be fine for all users. If anyone runs into a problem, we can advise them to delete failed payments first (we might need https://github.com/lightningnetwork/lnd/pull/5368 and a CLI tool for using that RPC as well) for that.
 2. Do the migration in four steps: one for each key (info/settle/fail), where we just copy the data and then a last step that deletes the attempt ID buckets. That way the individual transactions would still be atomic (meaning either all payments are copied or none are) but lnd will only start once all 4 steps were run. But if let's say step 3 fails, you're in a bad state as well, so this sounds like a bad idea...